### PR TITLE
feat(windows): Keyman Setup online bootstrap

### PR DESCRIPTION
**Pull request by mcdurdin**
This starts the refactoring of Keyman Setup for Windows for a more online-first bootstrap approach for bundling. For users who are offline, the traditional bundled approach works, as does throwing package files into the same folder as setup.exe. This makes the whole installer much simpler to deploy and test.

At this stage, there are a lot of rough edges, so need to refine these, including setup strings, language choices, and error and logging management. This will all come in a future PR, as this baseline should be working for the basic install case.

## Pull Request Set

This is part of a set of five Pull Requests. The recommended order of review is:

* keymanapp/downloads.keyman.com#29 (basic infrastructure to support custom filenames for Keyman Windows Bootstrap)
* keymanapp/api.keyman.com#83 (this PR APIs for the Windows bootstrap)
* keymanapp/keyman.com#120 (universal link landing page + Windows)
* keymanapp/keyman#3304 (this PR -- new Windows bootstrap)
* keymanapp/help.keyman.com#188 (documentation for Keyman Engine API changes)

# Screen Shots

## Setup main screen for `keyman-setup.khmer_angkor.km.exe`

image

The items that will be installed are shown on the bottom left. These are both going to be downloaded post-install. Clicking `Install Options` gives more detail and control of these.

## Install Options dialog

image

Note that the user cannot choose to install a keyboard without Keyman in this instance, because Keyman is not yet installed.

# TODO -- Upcoming Pull Requests

Note that some of these items may be addressed in website PRs.

### Functionality
- [ ] Download dialog should be integrated into the Setup dialog instead, with
      progress shown there. The downloads should all be done before the install
      starts. Errors in download of optional components should allow install to
      continue.
- [ ] MSI location and downloading
      - [ ] 3. downloading msi vs downloading full bundled installer,
      - [ ] 4. downloading an updated setup.exe -- how to handle searching for local packages, passing of command line parameters, etc.
      - [ ] 14. Look in current folder for .msi / .exe as well
- [ ] 17. Allow specification of tier in setup (-t for parameter; keyman-setup-alpha, keyman-setup-beta, keyman-setup[-stable] for embedded).
- [ ] 9. Embed MIT license
- [ ] 10. Consider mocks and unit tests
- [ ] 7. Handling offline scenarios and transitioning to online (e.g. allowing through firewall after prompt; retry needed?)
      - [ ] 16. Need to show 'checking online' dialog as part of this (except in silent mode)
- [ ] 15. Review unattended install scenarios, offline, online, etc.
- [ ] Online update checks need to support package deprecations
- [ ] Online update checks need to support major version upgrades
- [ ] Online update checks need to support /windows/14.0/update path
- [ ] Strip off " (1)" suffix that IE and other browsers add to filenames

### Cleanup
- [ ] 5. Cleanup of global variable ExtPath, TempPath, ProgramPath (use FInstallInfo)
- [ ] 1. localised strings (and management)
- [ ] 2. error management and logging (sentry?)
- [ ] Remove unused code:
  * `IsSameMajorVersion` https://github.com/keymanapp/api.keyman.com/pull/83#discussion_r449410956
  * `BuildKeyboardDownloadPath $version` parameter https://github.com/keymanapp/api.keyman.com/pull/83#discussion_r449412411

### Refactoring
- [ ] 8. Refactor of PackagesOnly vs ShouldInstallKeyman (PackagesOnly is awkward and needs to go)
      - [ ] 16. Consolidate big piles of boolean parameters into record (TInstallInfo.Options?)
- [ ] 11. Constants for GUIDs and file extensions etc
- [ ] 12. Refactor dependencies for ResourceDownloader to avoid circular dependencies

### Documentation
- [ ] 13. Update kmcomapi docs and kmshell -i docs.

### Bugs
- [ ] https://downloads.keyman.com/api/version/windows/2.0 is returning "exe" type for .msi files (probably a CI build script issue)

### CODE REVIEW

File: windows/src/desktop/kmshell/install/UfrmInstallKeyboard.pas [hunk 1]
```diff
@@ -161,9 +185,29 @@ function InstallFiles(Owner: TComponent; const FileNames: TWideStrings; ASilent:
     try
       kmcom.Keyboards.Refresh;
 
-      if AnsiSameText(ExtractFileExt(FileNames[i]), '.kmp')
-        then kmcom.Packages.Install(FileNames[i], True)
-        else kmcom.Keyboards.Install(FileNames[i], True);
+      FilenameBCP47 := Filenames[I].Split(['=']);
+      Filename := FilenameBCP47[0];
+      IsPackage := AnsiSameText(ExtractFileExt(FileNames[i]), '.kmp');
```
Comment (jahorton) [inline windows/src/desktop/kmshell/install/UfrmInstallKeyboard.pas:190]: I take it Delphi is case-insensitive?  I don't see a declaration for `i`, but I do for `I`.  If so, it's probably still better that we case variables consistently.

Since `i` is used through much of the rest of the function, perhaps it's the `I` declaration (line 183) that should shift.  Either way.

File: windows/src/desktop/kmshell/install/UfrmInstallKeyboard.pas [hunk 2]
```diff
@@ -161,9 +185,29 @@ function InstallFiles(Owner: TComponent; const FileNames: TWideStrings; ASilent:
...
+
+      if (Length(FilenameBCP47) > 1) and (FilenameBCP47[1] <> '') then
+      begin
+        if IsPackage then
+        begin
+          FPackage := (kmcom.Packages as IKeymanPackagesInstalled2).Install2(FileNames[i], True, False);
+          InstallKeyboardPackageLanguage(FPackage, FilenameBCP47[2]);
```
Comment (jahorton) [inline windows/src/desktop/kmshell/install/UfrmInstallKeyboard.pas:197]: Using your example of `keyman-setup.khmer_angkor.km.exe` as a filename, what do we expect to see here?   I'm trying to compare it against your example (`keyman-setup.khmer_angkor.km.exe`), but I'm coming up short.  (In particular, because of the `Filenames[I].Split(['=']);` line above; it's not clear where a `=` comes into play at all.

File: windows/src/desktop/setup/UfrmDownloadProgress.pas [hunk 3]
```diff
@@ -26,117 +25,70 @@
...
-  else inherited;
-  end;
+  Cancel := FCancel;
 end;
 
-procedure TfrmDownloadProgress.FormShow(Sender: TObject);
+procedure TfrmDownloadProgress.HTTPStatus(Sender: TObject;
+  const Message: string; Position, Total: Int64);  // I2855
 begin
-  inherited;
```
Comment (jahorton) [inline windows/src/desktop/setup/UfrmDownloadProgress.pas:138]: I see that this keyword was dropped when this method was repositioned within its source?

File: windows/src/engine/tsysinfox64/tsysinfox64.dproj [hunk 4]
```diff
@@ -118,6 +118,7 @@
         <DCC_DebugDCUs>true</DCC_DebugDCUs>
         <VerInfo_Keys>CompanyName=;FileVersion=5.0.0.26;InternalName=;LegalCopyright=;LegalTrademarks=;OriginalFilename=;ProductVersion=1.0.0.0;Comments=;ProgramID=com.embarcadero.$(MSBuildProjectName);FileDescription=$(MSBuildProjectName);ProductName=$(MSBuildProjectName)</VerInfo_Keys>
         <VerInfo_Locale>1033</VerInfo_Locale>
+        <BT_BuildType>Debug</BT_BuildType>
```
Comment (jahorton) [inline windows/src/engine/tsysinfox64/tsysinfox64.dproj:121]: Is this change intended?

File: windows/src/desktop/setup/Keyman.Setup.System.OnlineResourceCheck.pas [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+    url.AddParameter('package_'+pack.ID, pack.Locations.LatestVersion);
+  http := THTTPClient.Create;
+  try
+    response := http.Get(url.ToString);
+    if response.StatusCode <> 200 then
+    begin
+      // TODO: log failed response
+      //raise EOnlineUpdateCheck.Create('Error '+IntToStr(Response.StatusCode));
+      Exit;
+    end;
```
Comment (jahorton) [inline windows/src/desktop/setup/Keyman.Setup.System.OnlineResourceCheck.pas:61]: I assume you're noting this for implementation in the near future in some manner?

File: windows/src/desktop/setup/Keyman.Setup.System.OnlineResourceCheck.pas [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+
+    u := UTF8Encode(Response.ContentAsString(TEncoding.UTF8));
```
Comment (jahorton) [inline windows/src/desktop/setup/Keyman.Setup.System.OnlineResourceCheck.pas:63]: Mucky casing (`response` vs `Response`)

File: windows/src/desktop/setup/Keyman.Setup.System.ResourceDownloader.pas [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+    frmDownloadProgress.Callback := DownloadFileCallback;
+    Result := frmDownloadProgress.ShowModal = mrOk;
+  finally
+    frmDownloadProgress.Free;
+  end;
+end;
+
+procedure TResourceDownloader.HttpReceiveData(const Sender: TObject; AContentLength: Int64; AReadCount: Int64; var Abort: Boolean);
+begin
+  // TODO: stop using this form and report back to main form instead
```
Comment (jahorton) [inline windows/src/desktop/setup/Keyman.Setup.System.ResourceDownloader.pas:90]: Definitely looks like something to be deferred 'til later.  Worth making an issue for it, or does something related already exist?

File: windows/src/desktop/setup/RunTools.pas [hunk 8]
```diff
@@ -630,6 +506,7 @@ function TRunTools.InstallMSI: Boolean;
       // because we've backed up the relevant keys for reapplication post-install
       // Version 11 and later do not need this treatment as they are upgraded in-place
       // with the file and registry locations remaining static
+      // TODO: make these codes constants somewhere
```
Comment (jahorton) [inline windows/src/desktop/setup/RunTools.pas:509]: Another `TODO:`.

File: windows/src/desktop/setup/bootstrapmain.pas [hunk 9]
```diff
@@ -102,30 +72,267 @@ implementation
...
+  end;
+  SetExitVal(ERROR_OLD_WIN_VERSION);
+  Exit(True);
+end;
+
+// TODO: move this into a separate unit: it deals with installing keyboards into
+// an existing Keyman install on downlevel versions of Windows (e.g. Vista). So
+// we are constrained in how we do this -- no ability to do language associations
+// etc.
+procedure InstallKeyboardsInOldVersion(const ShellPath: string);   // I4460
```
Comment (jahorton) [inline windows/src/desktop/setup/bootstrapmain.pas:277]: Tracked issue?

File: windows/src/desktop/kmshell/install/UfrmInstallKeyboard.pas (hunk 1 above)
Comment (mcdurdin) [reply windows/src/desktop/kmshell/install/UfrmInstallKeyboard.pas:190]: Yes, Delphi is case-insensitive.

File: windows/src/desktop/kmshell/install/UfrmInstallKeyboard.pas (hunk 2 above)
Comment (mcdurdin) [reply windows/src/desktop/kmshell/install/UfrmInstallKeyboard.pas:197]: So this is parsing a command-line parameter for kmshell.exe, not the bootstrap. This uses the pattern `-i package_filename[=bcp47] [package_filename[=bcp47]] ...`

For example: `kmshell -i c:\temp\khmer_angkor.kmp=km c:\temp\sil_euro_latin.kmp=fr`

The bootstrap installer will parse its filename (which cannot safely use `=`, hence the `.` strategy), download the .kmp files, and pass the parameters into Keyman Configuration (kmshell).

In some ways it'd be nicer for the two strategies to be more consistent in their patterns, but I don't want to constrain the parameter model to the limited character set of the filename model...

(Note that the bootstrap installer *also* accepts a `-p package_id=bcp47[,package_id=bcp47]` parameter, which works the same way as embedding into the bootstrap filename but is more nicerer for running from command line).

File: windows/src/desktop/setup/UfrmDownloadProgress.pas (hunk 3 above)
Comment (mcdurdin) [reply windows/src/desktop/setup/UfrmDownloadProgress.pas:138]: Yes, the form is no longer an inherited form (as it was, a long time ago), so no need to call base methods (there aren't any!)

File: windows/src/engine/tsysinfox64/tsysinfox64.dproj (hunk 4 above)
Comment (mcdurdin) [reply windows/src/engine/tsysinfox64/tsysinfox64.dproj:121]: It was added by Delphi when I saved the group project. It is harmless (we always specify the build type directly during builds).

File: windows/src/desktop/setup/Keyman.Setup.System.OnlineResourceCheck.pas (hunk 5 above)
Comment (mcdurdin) [reply windows/src/desktop/setup/Keyman.Setup.System.OnlineResourceCheck.pas:61]: Yes, all TODO items are noted by me (didn't want to create issues for each one of them -- it's just too much noise).

File: windows/src/desktop/setup/Keyman.Setup.System.ResourceDownloader.pas (hunk 7 above)
Comment (mcdurdin) [reply windows/src/desktop/setup/Keyman.Setup.System.ResourceDownloader.pas:90]: Yes, noted (not as an issue).

File: windows/src/desktop/setup/RunTools.pas (hunk 8 above)
Comment (mcdurdin) [reply windows/src/desktop/setup/RunTools.pas:509]: Yes, noted (not as an issue).

File: windows/src/desktop/setup/bootstrapmain.pas (hunk 9 above)
Comment (mcdurdin) [reply windows/src/desktop/setup/bootstrapmain.pas:277]: Yes, noted (not as an issue).

File: windows/src/desktop/kmshell/install/UfrmInstallKeyboard.pas (hunk 2 above)
Comment (jahorton) [reply windows/src/desktop/kmshell/install/UfrmInstallKeyboard.pas:197]: Thanks, that's the missing information I was looking for.  Makes a lot more sense now.  Might be useful to have this in comments / in-code documentation somewhere.
